compiler/rustc_privacy/src/lib.rs

```diff
@@ -1309,6 +1309,23 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
         let def = match qpath {
             hir::QPath::Resolved(_, path) => match path.res {
                 Res::Def(kind, def_id) => Some((kind, def_id)),
+                // HACK: prevent using the `Self` ctor when the concrete type is private.
+                Res::SelfCtor(impl_) => {
```

Comment:
it's very annoying, but we normalize before looking up the constructor iirc, so the following should still be an issue:
```rust
mod b {
    #[derive(Default)]
    pub struct A(u32);
}

pub trait B {
    fn f(&self);
}

pub trait Id {
    type Assoc;
}
impl<T> Id for T {
    type Assoc = Self;
}

impl B for <b::A as Id>::Assoc {
    fn f(&self) {
        let Self(a) = self;
        //~^ ERROR: tuple struct constructor `A` is private
        println!("{}", a);
    }
}

fn main() {
    let a = b::A::default();
    a.f();
}
```
while ugly, it might be safer to do the check in astconv or sth :thinking: I am not familiar with this code

Reply:
It doesn't allow projections:

```
error: the `Self` constructor can only be used with tuple or unit structs
  --> /home/gh-fee1-dead/rust/tests/ui/privacy/issue-111220-tuple-struct-fields.rs:28:13
   |
LL |         let Self(a) = self;
   |             ^^^^^^^

error[E0164]: expected tuple struct or tuple variant, found self constructor `Self`
  --> /home/gh-fee1-dead/rust/tests/ui/privacy/issue-111220-tuple-struct-fields.rs:28:13
   |
LL |         let Self(a) = self;
   |             ^^^^^^^ not a tuple struct or tuple variant
```

Reply:
what? :sweat_smile: this compiles on nightly: https://rust.godbolt.org/z/nM6vzrc5Y

how does your code change this '^^

Reply:
Ahh, I turned `b::A::Assoc` into `(b::A,)` which isn't a tuple struct lol. This is indeed an issue, I will look into this when I get time.

Reply:
I used `try_normalize_erasing_regions` which should fix this for `impl Trait for Foo::AssocTy`. Let me know if this should totally be at some other crate and not `rustc_privacy`. I mainly put this stuff here so I can use `item_is_accessible`.

Reply:
I think normalizing it with the impl where bounds is still not enough as you can avoid it as follows (please add this as a test)
```rust
mod b {
    pub struct A(u32);
}

trait Id {
    type Assoc;
}
impl Id for b::A {
    type Assoc = b::A;
}
impl Id for u32 {
    type Assoc = u32;
}


trait Trait<T> {
    fn method(&self)
    where
        T: Id<Assoc = b::A>;
}

impl<T: Id> Trait<T> for <T as Id>::Assoc {
    fn method(&self)
    where
        T: Id<Assoc = b::A>,
    {
        let Self(a) = self;
        println!("{a}");
    }
}
```
You would have to normalize in the `param_env` of `self.current_item` I think but even then it feels somewhat dangerous.

Can you add a check to `instantiate_value_path` that the used ctor for is actually accessible :thinking: either completely move this check to there or at least `delay_span_bug` if we expect it to error.

Reply:
Yeah, I think this could be moved to typeck.